### ISSUE TITLE
Unable to generate iOS predicate for accessing an element using Appium with Java

### ISSUE THREAD

**Issue by surikiran** [appium/appium#7837]
## The problem

I am planning to access an element using predicate for which I found the XPath using Appium inspector. After going through the tutorials I tried to design a predicate but for my bad luck not even a single thing worked. I went through the forums too but unable to find a solution as this is the first ever time I started using predicates as XPaths are creating issues and seems to be not feasible for iOS Mobile Automation with Appium as suggested by a lot and few issues which I have been facing

## Environment

Appium 1.5.3
Max OS X 10.11.6
iphone 5s(8.2) and iPad 6 Plus(8.3)

## Details
I want to click a label contained among a list of a table in a screen of iOS Native App for which I had been using XPath **/UIAApplication[1]/UIAWindow[1]/UIAScrollView[1]/UIATableView[1]/UIATableCell[1]/UIATableView[2]/UIATableCell[2] which didn't serve my purpose due to the slow execution of XPath resulted in other blockers due to which I found that XPath is not useful hence understood from different forums that predicates is the best solution. I used to get XPath for my elements from Appium inspector but I found there is no such I used my XPath values to frame a predicate but this predicate didn't work besides I have been getting the error org.openqa.selenium.NoSuchElementException: An element could not be located on the page using the given search parameters. as I understood that my predicate is not identified with the line driver.findElementByIosUIAutomation (".tableViews()[1].cells().firstWithPredicate(\"name LIKE 'UC 6302'\")"). I used all MATCHES,ANY,CONTAINS but not even single one worked hence need help to solve this issue.

## Link to Appium logs

org.openqa.selenium.NoSuchElementException: An element could not be located on the page using the given search parameters. (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 20.74 seconds
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:32:46'
System info: host: 'Product-QA-Macmini.local', ip: '172.17.12.180', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.11.6', java.version: '1.8.0_45'
Driver info: io.appium.java_client.ios.IOSDriver
Session ID: cf22ae03-8e7a-4cd6-b6a3-fc892f755fc0
*** Element info: {Using=-ios uiautomation, value=.tableViews()[1].cells().firstWithPredicate("name LIKE 'UC 6302'")}
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:678)
	at io.appium.java_client.DefaultGenericMobileDriver.execute(DefaultGenericMobileDriver.java:40)
	at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:1)
	at io.appium.java_client.ios.IOSDriver.execute(IOSDriver.java:1)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:363)
	at io.appium.java_client.DefaultGenericMobileDriver.findElement(DefaultGenericMobileDriver.java:56)
	at io.appium.java_client.AppiumDriver.findElement(AppiumDriver.java:1)
	at io.appium.java_client.ios.IOSDriver.findElement(IOSDriver.java:1)
	at io.appium.java_client.ios.IOSDriver.findElementByIosUIAutomation(IOSDriver.java:207)
	at com.mobienxV6_3.automation.usecases.UC6302_Mobile.UC6302ScreensNavigation(UC6302_Mobile.java:450)
	at com.mobienxV6_3.automation.usecases.UC6302_Mobile.UC6302Menu(UC6302_Mobile.java:389)
	at com.mobienxV6_3.automation.execution.iOSDeviceExecution.MainMenu(iOSDeviceExecution.java:200)
	at com.mobienxV6_3.automation.execution.iOSDeviceExecution.ExecutionMode(iOSDeviceExecution.java:111)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:100)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:216)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:143)
	at org.testng.TestRunner.beforeRun(TestRunner.java:631)
	at org.testng.TestRunner.run(TestRunner.java:599)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:368)
	at org.testng.SuiteRunner.access$000(SuiteRunner.java:39)
	at org.testng.SuiteRunner$SuiteWorker.run(SuiteRunner.java:402)
	at org.testng.internal.thread.ThreadUtil$2.call(ThreadUtil.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

## Code To Reproduce Issue [ Good To Have ]

driver.findElementByIosUIAutomation (".tableViews()[1].cells().firstWithPredicate(\"name LIKE 'UC 6302'\")").

**Comment by imurchie** [comment]
See our documentation, and the links to Apple's, https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/ios_predicate.md#ios-predicate

Have you tried escaping the string?

```java
driver.findElementByIosUIAutomation (".tableViews()[1].cells().firstWithPredicate(\"name LIKE 'UC 6302'\")");
```

**Comment by surikiran** [comment]
@imurchie I tried as per your suggestion and this time surprisingly I got the below error:-
**Modified Code**
driver.findElementsByIosUIAutomation (".tableViews()[1].cells().firstWithPredicate('name LIKE 'UC 6302'')" );

**Error Generated**
org.openqa.selenium.WebDriverException: Unexpected identifier 'UC'. Expected ')' to end a argument list. (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 863 milliseconds
*** Element info: {Using=-ios uiautomation, value=.tableViews()[1].cells().firstWithPredicate('name LIKE 'UC 6302'')}
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:678)
	at io.appium.java_client.DefaultGenericMobileDriver.execute(DefaultGenericMobileDriver.java:40)
	at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:1)
	at io.appium.java_client.ios.IOSDriver.execute(IOSDriver.java:1)
	at org.openqa.selenium.remote.RemoteWebDriver.findElements(RemoteWebDriver.java:388)
	at io.appium.java_client.DefaultGenericMobileDriver.findElements(DefaultGenericMobileDriver.java:48)
	at io.appium.java_client.AppiumDriver.findElements(AppiumDriver.java:195)
	at io.appium.java_client.ios.IOSDriver.findElementsByIosUIAutomation(IOSDriver.java:216)
	at com.mobienxV6_3.automation.utils.WaitScenarios.waitByIOSAutomation(WaitScenarios.java:44)
	at com.mobienxV6_3.automation.usecases.UC6302_Mobile.UC6302ScreensNavigation(UC6302_Mobile.java:450)
	at com.mobienxV6_3.automation.usecases.UC6302_Mobile.UC6302Menu(UC6302_Mobile.java:389)
	at com.mobienxV6_3.automation.execution.iOSDeviceExecution.MainMenu(iOSDeviceExecution.java:200)
	at com.mobienxV6_3.automation.execution.iOSDeviceExecution.ExecutionMode(iOSDeviceExecution.java:111)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:100)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:216)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:143)
	at org.testng.TestRunner.beforeRun(TestRunner.java:631)
	at org.testng.TestRunner.run(TestRunner.java:599)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:368)
	at org.testng.SuiteRunner.access$000(SuiteRunner.java:39)
	at org.testng.SuiteRunner$SuiteWorker.run(SuiteRunner.java:402)
	at org.testng.internal.thread.ThreadUtil$2.call(ThreadUtil.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
I have used the escape character after getting this error prior to raising this issue. I concluded using the escape characted '\' going through the tutorials as per the link provided by you. I need the string 'UC 6302' as it is the name of the label as designed by developer. When tried with XPath I never got this type issues apart from performance due to which the app gets crashed but as per your suggestion strangely it didn't pass even the first step hence can you please suggest what more changes can be done to get the correct predicate...

**Comment by imurchie** [comment]
I don't know what part of my suggestion you tried, but `driver.findElementsByIosUIAutomation (".tableViews()[1].cells().firstWithPredicate('name LIKE 'UC 6302'')" );` was not something that is useful.

You need to constuct the string such that it can be parsed as code.

```
".tableViews()[1].cells().firstWithPredicate('name LIKE 'UC 6302'')"
```
Will not be parsed as a valid code. You need to make sure you don't cut things off in the middle, using escaping, since you have embedded quotes.
```
".tableViews()[1].cells().firstWithPredicate(\"name LIKE 'UC 6302'\")"
```

**Comment by surikiran** [comment]
Hi @imurchie I am unable to undestand what you said...? Can you please explain what do you mean by "...using escaping....". I am trying to understand that you are asking me to remove "\" and use the predicate without this symbol....? Can you please help me what shall I do to make this statement executed successfully.As this is the first time in using predicates besides having very less knowledge in programming I am unable to understand your answer. As suggested by you in my statement 
driver.findElementsByIosUIAutomation (".tableViews()[1].cells().firstWithPredicate('name LIKE 'UC 6302'')" );
what shall I need to remove to get a successful execution in the above statement.......please help....

**Comment by imurchie** [comment]
The string passed into this method is treated as code on the device. So if you pass in
```
".tableViews()[1].cells().firstWithPredicate('name LIKE 'UC 6302'')"
```
It is treated as:
```
.tableViews()[1].cells().firstWithPredicate('name LIKE 'UC 6302'')
```
The device, then understands 
```
.tableViews()[1].cells().firstWithPredicate('name LIKE '
```
because the string is finished, and it then cannot figure out what the rest (`UC 6302'')`) is doing. 

So you need to get it so that you have _one_ string as a parameter to `firstWithPredicate`. Which means you need to escape some quotation marks.
```
'name LIKE 'UC 6302''
```
is a string (`'name LIKE '`), some random stuff (`UC 6302`) and an empty string (`''`).

By escaping the internal quotation marks, this is a single string
```
'name LIKE \'UC 6302\''
```

See https://en.wikipedia.org/wiki/Escape_character

**Comment by surikiran** [comment]
Hi @imurchie thanks a lot for your help.....actually not exactly the same but your description helped to identify my mistake and after going through the forums and tutorials again and again I finally came up with the successful string ".scrollViews()[0].tableViews()[0].cells()[0].tableViews()[1].cells()[1]" as the escape characters and the existing string as per the application are creating few other issues making me more confused I found that providing the index values of the array is bit easy for getting the exact predicate. Once again thanks a lot for your help. I have one more issue with the predicate. Shall we close this issue and raise and fresh issue for that else shall I mention the same here. Based on your confirmation we can conclude whether to close this ticket/issue.
